prow/jira/jira.go

```diff
@@ -259,8 +288,275 @@ func (jc *client) UpdateRemoteLink(id string, link *jira.RemoteLink) error {
+		return fmt.Errorf("No transition status with name `%s` could be found. Please select from the following list: %v", statusName, nameList)
+	}
+	return jc.DoTransition(issueID, transitionID)
+}
+
+// UpdateStatus updates an issue's status by identifying the ID of the provided
+// statusName and then doing the status transition to update the issue.
+func (jc *client) UpdateStatus(issueID, statusName string) error {
+	return GenericUpdateStatus(jc, issueID, statusName)
+}
+
+func (jc *client) GetTransitions(issueID string) ([]jira.Transition, error) {
+	transitions, resp, err := jc.upstream.Issue.GetTransitions(issueID)
+	if err != nil {
+		return nil, JiraError(resp, err)
+	}
+	return transitions, nil
+}
+
+func (jc *client) DoTransition(issueID, transitionID string) error {
+	resp, err := jc.upstream.Issue.DoTransition(issueID, transitionID)
+	if err != nil {
+		return JiraError(resp, err)
+	}
+	return nil
+}
+
+func (jc *client) UpdateIssue(issue *jira.Issue) (*jira.Issue, error) {
+	result, resp, err := jc.upstream.Issue.Update(issue)
+	if err != nil {
+		return nil, JiraError(resp, err)
+	}
+	return result, nil
+}
+
+func (jc *client) AddComment(issueID string, comment *jira.Comment) (*jira.Comment, error) {
+	result, resp, err := jc.upstream.Issue.AddComment(issueID, comment)
+	if err != nil {
+		return nil, JiraError(resp, err)
+	}
+	return result, nil
+}
+
+func (jc *client) CreateIssue(issue *jira.Issue) (*jira.Issue, error) {
+	result, resp, err := jc.upstream.Issue.Create(issue)
+	if err != nil {
+		return nil, JiraError(resp, err)
+	}
+	return result, nil
+}
+
+func (jc *client) CreateIssueLink(link *jira.IssueLink) error {
+	resp, err := jc.upstream.Issue.AddLink(link)
+	if err != nil {
+		return JiraError(resp, err)
+	}
+	return nil
+}
+
+// GenericCloneIssue is a generic implementation of CloneIssue that can be used with
+// any valid Client.
+func GenericCloneIssue(jc Client, parent *jira.Issue) (*jira.Issue, error) {
+	// create deep copy of parent "Fields" field
+	data, err := json.Marshal(parent.Fields)
+	if err != nil {
+		return nil, err
+	}
+	childIssueFields := &jira.IssueFields{}
+	err = json.Unmarshal(data, childIssueFields)
+	if err != nil {
+		return nil, err
+	}
+	childIssue := &jira.Issue{
+		Fields: childIssueFields,
+	}
+	// update description
+	childIssue.Fields.Description = fmt.Sprintf("This is a clone of issue %s. The following is the description of the original issue: \n---\n%s", parent.Key, parent.Fields.Description)
+
+	// attempt to create the new issue
+	createdIssue, err := jc.CreateIssue(childIssue)
+	if err != nil {
+		// some fields cannot be set on creation; unset them
+		childIssue, err = unsetProblematicFields(childIssue, err)
+		if err != nil {
+			return nil, fmt.Errorf("error identifying unsettable fields during clone creation: %w", err)
+		}
+		createdIssue, err = jc.CreateIssue(childIssue)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// create clone links
+	link := &jira.IssueLink{
```

Comment:
There is an `IssueLinks` field on `Issue.Fields`, why not put the link there prior to creating it and avoid the update call and following fetch?

Reply:
`IssueLinks` cannot be set on issue creation or update. Am issue link must be created via the issue links API instead.